File: pet-store-service/src/main/java/com/chewy/osn/exception/PetExceptionHandler.java
package com.chewy.osn.exception;

import io.micronaut.http.HttpResponse;
import io.micronaut.http.HttpStatus;
import io.micronaut.http.MediaType;
import io.micronaut.http.MutableHttpResponse;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Error;
import io.micronaut.http.simple.SimpleHttpResponseFactory;
import jakarta.inject.Singleton;

@Controller
@Singleton
public class PetExceptionHandler {

    @Error(global = true)
    public HttpResponse<HttpErrorBody> handle(PetException exception) {
        HttpStatus status = HttpStatus.valueOf(exception.getCode());
        MutableHttpResponse<HttpErrorBody> response = SimpleHttpResponseFactory.INSTANCE.status(status);
        response.contentType(MediaType.APPLICATION_JSON_TYPE);
        response.body(buildResponse(exception));
        return response;
    }

    private HttpErrorBody buildResponse(PetException exception) {
        HttpErrorBody response = new HttpErrorBody();
        response.setMessage(exception.getMessage());
        return response;
    }
}

File: pet-store-service/src/main/kotlin/com/johnkruss/osn/exception/PetExceptionHandler.kt
package com.johnkruss.osn.exception

import io.micronaut.http.HttpResponse
import io.micronaut.http.HttpStatus
import io.micronaut.http.MediaType
import io.micronaut.http.annotation.Controller
import io.micronaut.http.annotation.Error
import io.micronaut.http.simple.SimpleHttpResponseFactory
import jakarta.inject.Singleton

@Controller
@Singleton
class PetExceptionHandler {
    @Error(global = true)
    fun handle(exception: PetException): HttpResponse<HttpErrorBody> {
        val status = HttpStatus.valueOf(exception.code)
        val response = SimpleHttpResponseFactory.INSTANCE.status<HttpErrorBody>(status)
        response.contentType(MediaType.APPLICATION_JSON_TYPE)
        response.body(buildResponse(exception))
        return response
    }

    private fun buildResponse(exception: PetException): HttpErrorBody {
        val response = HttpErrorBody(exception.message)
        return response
    }
}
